example-app: prevent redirect from LoginEndpoint when login fails

nagylzs opened the pull request:
Issue: #84

VasilMitovClouway commented:
Would be best if you squash your commits to one that will better describe the fix. An example: 
 oauth2/example-app: on login fail a redirect no longer occurs or the current title of the PR is a good candidate as well. We usually try to keep the policy of single commit per PR.

VasilMitovClouway commented on oauth2-example-app/src/main/java/com/clouway/oauth2/exampleapp/LoginEndpoint.java:
```diff
@@ -17,6 +20,7 @@
 public class LoginEndpoint {
   private String username;
   private String password;
+  private String errormessage;
```
use camelCase when naming - errorMessage

nagylzs replied:
It it not camelCase because the LoginEndpoint receives it as a get parameter. If I change it to errorMessage then the GET parameter must also be camelCase.

VasilMitovClouway commented on oauth2-example-app/src/main/java/com/clouway/oauth2/exampleapp/LoginEndpoint.java:
```diff
@@ -50,4 +57,12 @@ public void setPassword(String password) {
     this.password = password;
   }
 
+  public String getErrormessage() {
```
These are not used. Usually getters and setters are to be avoided. 
If you want more on the matter : https://dev.to/scottshipp/avoid-getters-and-setters-whenever-possible-c8m

VasilMitovClouway commented on oauth2-example-app/src/main/java/com/clouway/oauth2/exampleapp/storage/InMemoryModule.java:
```diff
@@ -36,7 +39,24 @@ protected void configure() {
     InMemoryClientRepository clientRepository = new InMemoryClientRepository();
-    clientRepository.register(new ClientRegistrationRequest("857613db7b18232c72a5093ad19dbc6df74a139e", "test", Collections.singleton("http://localhost:8080/oauth/callback")));
+    
+    String registeredRedirectURI = "https://httpbin.org/get";
```
Redirect URL looks strange.

nagylzs replied:
It is just for the example, it redirects to a page that displays all parameters on a page, so anyone can try example oauth server alone (without creating another server to be redirected to), without modifications.

nagylzs replied:
Changes removed, anyway

VasilMitovClouway commented on oauth2-example-app/src/main/java/com/clouway/oauth2/exampleapp/LoginEndpoint.java:
```diff
@@ -26,7 +30,7 @@ public LoginEndpoint(ResourceOwnerAuthentication authentication) {
 
   @Post
-  public String login(HttpServletRequest request, HttpServletResponse response) {
+  public String login(HttpServletRequest request, HttpServletResponse response) throws UnsupportedEncodingException {
```
Would be better if the error is handled and not thrown by the Post method.

nagylzs replied:
Okay should I throw an Error instead like this?

 catch(... e) {throw new Error("The world is ending, UTF-8 is not an encoding anymore!", e);}

or maybe I could use URLEncoder.encode without the encoding parameter (but that is deprecated API)

VasilMitovClouway commented on oauth2-example-app/src/main/java/com/clouway/oauth2/exampleapp/storage/InMemoryModule.java:
```diff
@@ -36,7 +39,24 @@ protected void configure() {
+    String registeredRedirectURI = "https://httpbin.org/get";
+    Client client = clientRepository.register(new ClientRegistrationRequest("857613db7b18232c72a5093ad19dbc6df74a139e", "test", Collections.singleton(registeredRedirectURI)));
+	try {
+		System.err.println(
```
We do not have an error here.

nagylzs replied:
Changes removed

VasilMitovClouway commented on oauth2-example-app/src/main/java/com/clouway/oauth2/exampleapp/storage/InMemoryResourceOwnerRepository.java:
```diff
@@ -20,6 +20,12 @@
   private Map<String, ResourceOwner> resourceOwners = new HashMap<String, ResourceOwner>() {{
     put("admin", new ResourceOwner("admin", "admin"));
+    
+    System.err.println(
```
.

VasilMitovClouway replied:
I think this message is not needed.

nagylzs replied:
Message removed.

VasilMitovClouway commented on oauth2-example-app/src/test/java/com/clouway/oauth2/exampleapp/LoginEndpointTest.java:
```diff
@@ -13,6 +13,10 @@
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertNotNull;
```
Unused import.

VasilMitovClouway commented on oauth2-example-app/src/test/java/com/clouway/oauth2/exampleapp/LoginEndpointTest.java:
```diff
@@ -13,6 +13,10 @@
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import java.net.URLEncoder;
```
.

VasilMitovClouway commented on oauth2-example-app/src/main/java/com/clouway/oauth2/exampleapp/LoginEndpoint.java:
```diff
@@ -50,4 +57,12 @@ public void setPassword(String password) {
+	return errormessage;
+  }
+
+  public void setErrormessage(String errormessage) {
```
You are not posting the message from the login page. The username and password setters are used by sitebricks to set them but the error message is not posted so it is never set or get.

nagylzs replied:
The LoginPage redirects to itself with &errormessage=Invalid+username+or+password and that sets the field.

nagylzs commented:
Removed the unnecessary changes, removed unused imports, fixed the LoginEndpointTest. The tests are passing. The only change I did not make was the errorMessage camelcase and the setter, because it is needed to display the error message. I'll try to do everything in a single commit next time.

VasilMitovClouway commented on oauth2-example-app/src/test/java/com/clouway/oauth2/exampleapp/LoginEndpointTest.java:
```diff
@@ -59,6 +61,7 @@ public void wrongAuthentication() throws Exception {
       will(Expectations.returnValue(remoteAddress));
       oneOf(request).getParameter("continue");
       will(returnValue("/xxx"));
+      allowing(request).getRequestURI();
```
It's a bad idea to use allowing. This means that it can be called any number of times, and if you write 100 lines of request.getRequestURI the test will still pass. In the future if you wish to show that a method is called many times it would be better to specify how many times it is called with *exactly(n).of*. 
Here's a cheat sheet for jMock http://www.jmock.org/cheat-sheet.html

nagylzs replied:
Changed to OneOf

VasilMitovClouway commented:
We will wait for @mgenov to take a look at the PR as well. While we wait you could do from the console 
git rebase -i HEAD~[number of commits] to rebase the last  commits into one

nagylzs commented:
Squashed commits into one, force pushed
